DOCS: Add handling of GIL to pure.rst (#4650)
cython/cython: docs/src/tutorial/pure.rst

        return a == b

Managing the Global Interpreter Lock
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

* ``cython.nogil`` can be used as a context manager or as a decorator to replace the :keyword:`nogil` keyword::

    with cython.nogil:
        # code block with GIL released

    @cython.nogil
    @cython.cfunc
    def func_released_gil() -> cython.int:
        # function with GIL released

* ``cython.gil`` can be used as a context manager to replace the :keyword:`gil` keyword::

    with cython.gil:
        # code block with GIL acquired

>   .. Note:: Cython currently does not support ``@cython.gil`` decorator. ``@cython.nogil(False)`` should be used instead.

Reviewer: ```suggestion
  .. Note:: Cython currently does not support the ``@cython.gil`` decorator. ``@cython.nogil(False)`` should be used instead.
```

        return a == b

Managing the Global Interpreter Lock
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

* ``cython.nogil`` can be used as a context manager or as a decorator to replace the :keyword:`nogil` keyword::

    with cython.nogil:
        # code block with the GIL released

    @cython.nogil
    @cython.cfunc
    def func_released_gil() -> cython.int:
        # function with the GIL released

* ``cython.gil`` can be used as a context manager to replace the :keyword:`gil` keyword::

    with cython.gil:
        # code block with the GIL acquired

>   .. Note:: Cython currently does not support the ``@cython.gil`` decorator.

Reviewer 2: I think it should be called `@cython.with_gil`. Just naming it `@cython.gil` seems unclear. (We named the modifier "... with gil" for the same reason.)

Author: Me sense. I will update that part.

> When either ``cython.nogil`` or ``cython.gil`` is used as a function, it accepts a parameter for conditional
> releasing or acquiring the GIL. The condition must be constant (at compile time)::

Reviewer 2: ```suggestion
Both directives accept an optional boolean parameter for conditionally
releasing or acquiring the GIL. The condition must be constant (at compile time)::
```

Reviewer 2: It would also be nice to have an actual usage example for something less trivial. Passing `True` and `False` makes it look useless.

Author: > EDIT: or mention at least fused types as a common use case, with a type based condition.

Added

        return a == b

Managing the Global Interpreter Lock
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

* ``cython.nogil`` can be used as a context manager or as a decorator to replace the :keyword:`nogil` keyword::

    with cython.nogil:
        # code block with the GIL released

    @cython.nogil
    @cython.cfunc
    def func_released_gil() -> cython.int:
        # function with the GIL released

* ``cython.gil`` can be used as a context manager to replace the :keyword:`gil` keyword::

    with cython.gil:
        # code block with the GIL acquired

  .. Note:: Cython currently does not support the ``@cython.with_gil`` decorator.

Both directives accept an optional boolean parameter for conditionally
releasing or acquiring the GIL. The condition must be constant (at compile time)::

  with cython.nogil(False):
      # code block with the GIL not released

  @cython.nogil(True)
  @cython.cfunc
  def func_released_gil() -> cython.int:
      # function with the GIL released

  with cython.gil(False):
      # code block with the GIL not acquired

  with cython.gil(True):
      # code block with the GIL acquired

> The common use case of conditional acquiring and releasing of the GIL are fused types (see :ref:`gil_conditional`).

Reviewer 2: ```suggestion
A common use case for conditionally acquiring and releasing the GIL are fused types
that allow different GIL handling depending on the specific type (see :ref:`gil_conditional`).
```
